File: backend/app/models/website_analysis.py
"""Tabla `website_analyses`: auditorías Playwright guardadas, ligadas a un lead."""

from __future__ import annotations

from datetime import datetime, timezone

from sqlmodel import Field, SQLModel


def _utc_now() -> datetime:
    # Timestamps siempre en UTC
    return datetime.now(timezone.utc)


class WebsiteAnalysis(SQLModel, table=True):
    """Playwright-powered website audit saved for a lead."""

    __tablename__ = "website_analyses"

    id: int | None = Field(default=None, primary_key=True)     # autoincremental
    lead_id: int = Field(foreign_key="leads.id", index=True)   # a qué lead pertenece
    url: str                                                   # URL solicitada
    final_url: str | None = None                               # URL final tras redirects
    page_title: str | None = None                              # <title> de la página
    last_modified: str | None = None                           # header Last-Modified
    load_time_ms: float | None = None                          # tiempo de carga medido
    dom_content_loaded_ms: float | None = None                 # DOMContentLoaded medido
    technologies_json: str | None = None                       # lista JSON del stack detectado
    seo_title: str | None = None                               # metadatos SEO extraídos ↓
    seo_meta_description: str | None = None
    seo_h1: str | None = None
    seo_issues_json: str | None = None                         # lista JSON de problemas SEO
    summary: str | None = None                                 # resumen legible
    metrics_json: str | None = None                            # dict JSON de métricas extra
    status: str = Field(default="completed", index=True)       # completed | failed
    error_message: str | None = None                           # detalle si falló
    created_at: datetime = Field(default_factory=_utc_now)     # cuándo se corrió

File: frontend-v2/app/types/website-analysis.ts
/** Historial del dominio en Internet Archive (Wayback Machine). */
export interface WaybackInfo {
  domain: string                        // dominio consultado
  first_seen: string | null             // primera captura conocida
  last_seen: string | null              // última captura
  snapshot_count: number                // nº de capturas archivadas
  age_years: number | null              // antigüedad estimada del sitio
  timeline_url: string                  // link a la línea de tiempo completa
  first_snapshot_url: string | null     // link a la captura más antigua
  last_snapshot_url: string | null      // link a la más reciente
  available: boolean                    // ¿hay capturas para este dominio?
}

/** Un resultado de la búsqueda DuckDuckGo del sitio del negocio. */
export interface WebsiteSearchResult {
  title: string                    // título del resultado
  url: string                      // URL candidata
  snippet: string                  // extracto del resultado
  wayback?: WaybackInfo | null     // info de archive.org si ya se consultó
}

/** Un análisis de sitio web hecho con Playwright (auditoría técnica/SEO). */
export interface WebsiteAnalysis {
  id: number
  lead_id: number                       // a qué lead pertenece
  url: string                           // URL analizada
  final_url: string | null              // URL final tras redirects
  page_title: string | null             // <title> de la página
  last_modified: string | null          // header Last-Modified (si existe)
  load_time_ms: number | null           // tiempo de carga medido
  dom_content_loaded_ms: number | null  // DOMContentLoaded medido
  technologies: string[]                // stack detectado (WordPress, React…)
  seo_title: string | null              // metadatos SEO extraídos
  seo_meta_description: string | null
  seo_h1: string | null
  seo_issues: string[]                  // problemas SEO encontrados
  summary: string | null                // resumen legible del análisis
  metrics: Record<string, unknown>      // métricas extra (incluye wayback)
  status: string                        // completed | failed | …
  error_message: string | null          // detalle si falló
  created_at: string
}

/** Estado global del analizador: solo un Playwright a la vez. */
export interface AnalysisStatus {
  busy: boolean              // ¿hay un análisis corriendo ahora?
  lead_id: number | null     // de qué lead
  url: string | null         // qué URL está analizando
}
